index.js

```diff
@@ -1,156 +1,149 @@
+    if (wait <= have) return ready()
+    waitFor(archive, wait, ready)
 
-    var results = {
-      key: key,
-      filename: filename,
-      op: op
+    function ready () {
+      var arch = /^\d+$/.test(query.version) ? archive.checkout(Number(query.version)) : archive
+      if (name[name.length - 1] === '/') ondirectory(arch, name, req, res, opts)
+      else onfile(arch, name, req, res)
     }
-    debug('parse() results', results)
-    return results
   }
 }
 
-function serveFeedOrArchive (req, res, archive, urlOpts) {
-  debug('serveFeedOrArchive', archive.key.toString('hex'))
-  var opts = { live: urlOpts.op === 'changes' }
-  var through = new stream.PassThrough()
-  var src = archive.metadata ? archive.list(opts) : archive.createReadStream(opts)
-  var timeout = TimeoutStream({
-    objectMode: true,
-    duration: 10000
-  }, () => {
-    onerror(404, res)
-    src.destroy()
+function onfile (archive, name, req, res) {
+  archive.stat(name, function (err, st) {
+    if (err) return onerror(res, 404, err)
+
+    if (st.isDirectory()) {
+      res.statusCode = 302
+      res.setHeader('Location', name + '/')
+      return
+    }
+
+    var r = req.headers.range && range(st.size, req.headers.range)[0]
+    res.setHeader('Accept-Ranges', 'bytes')
+    res.setHeader('Content-Type', mime.lookup(name))
+
+    if (r) {
+      res.statusCode = 206
+      res.setHeader('Content-Range', 'bytes ' + r.start + '-' + r.end + '/' + st.size)
+      res.setHeader('Content-Length', r.end - r.start + 1)
+    } else {
+      res.setHeader('Content-Length', st.size)
+    }
+
+    if (req.method === 'HEAD') return res.end()
+    pump(archive.createReadStream(name, r), res)
   })
+}
 
-  res.setHeader('Content-Type', 'application/json')
-  if (archive.metadata) return pump(src, timeout, ndjson.serialize(), through)
-  return pump(src, timeout, through2.obj(function (chunk, enc, cb) {
-    cb(null, chunk.toString())
-  }), ndjson.serialize(), through)
+function ondirectory (archive, name, req, res, opts) {
+  archive.stat(name + 'index.html', function (err) {
+    if (err) return ondirectoryindex(archive, name, req, res, opts)
+    onfile(archive, name + 'index.html', req, res)
+  })
 }
 
-function serveFile (req, res, archive, filename) {
-  debug('serveFile', archive.key.toString('hex'), 'filename', [filename])
+function ondirectoryindex (archive, name, req, res, opts) {
+  list(archive, name, function (err, entries) {
+    if (err) entries = []
+
+    var wait = archive.metadata ? archive.metadata.length + 1 : 0
+    var script = `
+      function liveUpdate () {
+        var xhr = new XMLHttpRequest()
+        xhr.open("GET", "${name}?wait=${wait}", true)
+        xhr.onload = function () {
+          document.open()
+          document.write(xhr.responseText)
+          document.close()
+        }
+        xhr.onerror = function () {
+          setTimeout(liveUpdate, 1000)
+        }
+        xhr.send(null)
+      }
+
+      liveUpdate()
+    `
```

Comment:
•ogd> jhand: it live reloads the page when the data changes
3:17 PM <•karissa> every 1 second
3:17 PM <•ogd> jhand: any time the wait url resolves actually
3:17 PM <•karissa> ah
3:18 PM <jhand> ah ok
3:18 PM <•ogd> also doesnt reload just repaints html
3:18 PM <•ogd> jhand: if youre on version 1, it will do ?wait=2
3:18 PM <•ogd> jhand: and the server wont return http response for wait=2 until 2 exists
3:18 PM <jhand> nice
3:18 PM <•ogd> kind of clever